packages/cache/src/cache.js

```diff
@@ -0,0 +1,51 @@
+// @ts-check
+import { E } from '@endo/far';
+import { M } from '@agoric/store';
+
+import { makeState } from './state.js';
+import { makeStoreCoordinator } from './store.js';
+
+/** @template T @typedef {import('@endo/far').ERef<T>} ERef */
+
+/**
+ * @param {ERef<import('./types').Coordinator>} [coordinator]
+ */
+export const makeCache = (coordinator = makeStoreCoordinator()) => {
+  /**
+   * The ground state for a cache key value is `undefined`.  It is impossible to
+   * distinguish a set value of `undefined` from an unset key
+   *
+   * @param {unknown} key the cache key (any key type acceptable to the cache)
+   * @param {object | string | bigint | null | undefined | number | ((oldValue: any) => ERef<unknown>)} [update] a
+   * function that returns a `newValue` or promise for `newValue`
+   * @param {Pattern[]} optGuardPattern don't update unless this pattern matches
+   */
+  const cache = (key, update, ...optGuardPattern) => {
+    if (!update) {
+      return E.get(E(coordinator).getRecentState(key)).value;
+    }
+    const guardPattern =
+      optGuardPattern.length > 0 ? optGuardPattern[0] : M.any();
```

Comment:
Does this need to be an array if we're only using the first one?

Reply:
Yes, otherwise there is no way to distinguish between `cache(key, val)` and `cache(key, val, undefined)`.